File: frontend/src/types/alert.ts
export type Severity = "DEBUG" | "INFO" | "WARNING" | "ERROR" | "CRITICAL";

export interface AlertRule {
  id: string;
  user_id: string;
  name: string;
  severity: Severity;
  threshold: number;
  time_window_seconds: number;
  match_pattern: string | null;
  notification_channels: string[];
  is_active: boolean;
  created_at: string;
}

export interface AlertRuleCreate {
  name: string;
  severity: Severity;
  threshold: number;
  time_window_seconds?: number;
  match_pattern?: string | null;
  notification_channels?: string[];
  is_active?: boolean;
}

export interface AlertRuleUpdate {
  name?: string;
  severity?: Severity;
  threshold?: number;
  time_window_seconds?: number;
  match_pattern?: string | null;
  notification_channels?: string[];
  is_active?: boolean;
}

export interface AlertHistory {
  id: string;
  rule_id: string;
  triggered_at: string;
  resolved_at: string | null;
  severity: string;
  details: Record<string, unknown> | null;
}

File: backend/src/schemas/alert.py
"""
backend/src/schemas/alert.py

Pydantic schemas for alert rules and alert history.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


# ─────────────────────────────────────────────
# Alert Rule Schemas
# ─────────────────────────────────────────────

class AlertRuleBase(BaseModel):
    """Base fields shared across alert rule schemas."""
    name: str = Field(..., min_length=1, max_length=255)
    severity: str = Field(..., pattern=r"^(DEBUG|INFO|WARNING|ERROR|CRITICAL)$")
    threshold: int = Field(..., ge=1)
    time_window_seconds: int = Field(default=300, ge=1)
    match_pattern: str | None = Field(default=None, max_length=500)
    notification_channels: list[str] = Field(default_factory=list)
    is_active: bool = True


class AlertRuleCreate(AlertRuleBase):
    """Schema for creating a new alert rule (request body)."""
    pass


class AlertRuleUpdate(BaseModel):
    """Schema for updating an existing alert rule (partial update)."""
    name: str | None = Field(default=None, min_length=1, max_length=255)
    severity: str | None = Field(default=None, pattern=r"^(DEBUG|INFO|WARNING|ERROR|CRITICAL)$")
    threshold: int | None = Field(default=None, ge=1)
    time_window_seconds: int | None = Field(default=None, ge=1)
    match_pattern: str | None = Field(default=None, max_length=500)
    notification_channels: list[str] | None = None
    is_active: bool | None = None


class AlertRuleResponse(AlertRuleBase):
    """Schema for alert rule response (includes DB fields)."""
    id: str
    user_id: str
    created_at: datetime

    class Config:
        from_attributes = True


# ─────────────────────────────────────────────
# Alert History Schemas
# ─────────────────────────────────────────────

class AlertHistoryResponse(BaseModel):
    """Schema for alert history (triggered alert) response."""
    id: str
    rule_id: str
    triggered_at: datetime
    resolved_at: datetime | None
    severity: str
    details: dict[str, Any] | None

    class Config:
        from_attributes = True


class AlertRuleWithHistoryResponse(AlertRuleResponse):
    """Schema for alert rule response including its history."""
    history: list[AlertHistoryResponse] = []

    class Config:
        from_attributes = True
